### ISSUE TITLE
Advanced Suspicion Detection Algorithms - Enhanced Behavior Analysis

### ISSUE THREAD

**Issue by onnwee** [subculture-collective/discord-spywatcher#27]
## 🎯 Objective

Enhance suspicion detection with advanced algorithms and machine learning to more accurately identify suspicious behavior patterns.

### Enhanced Detection Algorithms
- [ ] **Multi-Factor Suspicion Scoring**
  - Ghost score (presence vs engagement)
  - Lurker score (read-only activity)
  - Multi-client frequency
  - Message timing patterns
  - Reaction time consistency
  - Account age correlation
  - Role change patterns
  - Time-of-day anomalies

- [ ] **Behavioral Analysis**
  - Message pattern analysis
  - Activity spike detection
  - Dormancy period detection
  - Cross-channel behavior
  - Interaction network analysis
  - Content similarity detection

- [ ] **Anomaly Detection**
  - Statistical outlier detection
  - Deviation from user's own baseline
  - Deviation from guild baseline
  - Sudden behavior changes
  - Coordinated activity detection

### Machine Learning (Future)

- [ ] **ML Model Development**
  - Training data collection
  - Feature engineering
  - Model selection (Random Forest, Neural Network)
  - Model training pipeline
  - Model evaluation metrics
  - A/B testing framework

- [ ] **Automated Learning**
  - Continuous learning from feedback
  - Admin feedback loop (false positives)
  - Pattern discovery
  - Model retraining schedule

### Suspicion Categorization

- [ ] **Suspicion Types**
  - Bot-like behavior
  - Lurker/spy behavior
  - Multi-account coordination
  - Automated responses
  - Time-zone anomalies
  - Engagement farming

- [ ] **Confidence Levels**
  - High confidence (95%+)
  - Medium confidence (70-95%)
  - Low confidence (50-70%)
  - Require review (<50%)

### Reporting & Visualization

- [ ] **Suspicion Dashboard**
  - Suspicion score timeline
  - Contributing factors breakdown
  - Comparison to guild average
  - Historical trends
  - Similar users
  - Recommended actions

- [ ] **Alert System**
  - Real-time suspicion alerts
  - Configurable thresholds
  - Alert channels (Discord, email, dashboard)
  - Alert fatigue prevention
  - False positive reporting

### Enhanced Suspicion Score Algorithm
```typescript
interface SuspicionFactors {
  ghostScore: number;          // 0-100
  lurkerScore: number;         // 0-100
  multiClientFreq: number;     // 0-100
  timingAnomalies: number;     // 0-100
  accountAge: number;          // days
  activityPattern: number;     // 0-100
  contentSimilarity: number;   // 0-100
}

interface SuspicionResult {
  totalScore: number;
  confidence: 'high' | 'medium' | 'low';
  factors: SuspicionFactors;
  reasons: string[];
  recommendedAction: string;
}

export const calculateAdvancedSuspicion = async (
  userId: string,
  guildId: string
): Promise<SuspicionResult> => {
  const user = await getUserData(userId, guildId);
  const guildAvg = await getGuildAverages(guildId);
  
  const factors: SuspicionFactors = {
    // Ghost score: presence vs engagement
    ghostScore: calculateGhostScore(user),
    
    // Lurker score: read vs write activity
    lurkerScore: calculateLurkerScore(user),
    
    // Multi-client login frequency
    multiClientFreq: calculateMultiClientFrequency(user),
    
    // Message timing anomalies
    timingAnomalies: await calculateTimingAnomalies(user, guildAvg),
    
    // Account age factor (newer = more suspicious)
    accountAge: user.accountAgeDays,
    
    // Activity pattern regularity (bots are too regular)
    activityPattern: calculateActivityPattern(user),
    
    // Content similarity (copy-paste detection)
    contentSimilarity: await calculateContentSimilarity(user),
  };
  
  // Weighted suspicion score
  const weights = {
    ghostScore: 0.20,
    lurkerScore: 0.15,
    multiClientFreq: 0.20,
    timingAnomalies: 0.15,
    accountAge: 0.10,
    activityPattern: 0.10,
    contentSimilarity: 0.10,
  };
  
  const totalScore = Object.entries(factors).reduce((sum, [key, value]) => {
    if (key === 'accountAge') {
      // Account age: newer is more suspicious
      const ageScore = Math.max(0, 100 - (value / 365) * 100);
      return sum + ageScore * weights[key];
    }
    return sum + value * weights[key];
  }, 0);
  
  // Determine confidence level
  const confidence = totalScore > 80 ? 'high' :
                    totalScore > 60 ? 'medium' : 'low';
  
  // Generate human-readable reasons
  const reasons: string[] = [];
  if (factors.ghostScore > 70) {
    reasons.push(`High presence (${user.presenceCount}) but low engagement (${user.messageCount} messages)`);
  }
  if (factors.multiClientFreq > 70) {
    reasons.push(`Frequently online from multiple devices simultaneously`);
  }
  if (factors.timingAnomalies > 70) {
    reasons.push(`Unusual timing patterns in messages and reactions`);
  }
  if (factors.activityPattern > 70) {
    reasons.push(`Suspiciously regular activity patterns (bot-like)`);
  }
  if (factors.accountAge < 30) {
    reasons.push(`New account (${factors.accountAge} days old)`);
  }
  
  // Recommended action
  const recommendedAction = 
    totalScore > 90 ? 'Consider banning - very high suspicion' :
    totalScore > 75 ? 'Monitor closely - high suspicion' :
    totalScore > 60 ? 'Investigate further - moderate suspicion' :
    'No immediate action needed';
  
  return {
    totalScore: Math.round(totalScore),
    confidence,
    factors,
    reasons,
    recommendedAction,
  };
};

// Calculate timing anomalies
const calculateTimingAnomalies = async (
  user: UserData,
  guildAvg: GuildAverages
): Promise<number> => {
  // Get reaction times for this user
  const reactionTimes = await db.reactionTime.findMany({
    where: { observerId: user.discordId },
    orderBy: { createdAt: 'desc' },
    take: 100,
  });
  
  if (reactionTimes.length < 10) return 0;
  
  // Calculate average and standard deviation
  const times = reactionTimes.map(rt => rt.deltaMs);
  const avg = times.reduce((a, b) => a + b, 0) / times.length;
  const stdDev = Math.sqrt(
    times.reduce((sq, n) => sq + Math.pow(n - avg, 2), 0) / times.length
  );
  
  // Very low standard deviation = bot-like
  // Times under 500ms consistently = suspicious
  const lowStdDevScore = stdDev < 100 ? 100 : (1000 - stdDev) / 10;
  const fastReactionScore = avg < 500 ? (500 - avg) / 5 : 0;
  
  // Compare to guild average
  const deviationFromNorm = Math.abs(avg - guildAvg.reactionTime) / guildAvg.reactionTime * 100;
  
  return Math.min(100, (lowStdDevScore + fastReactionScore + deviationFromNorm) / 3);
};

// Activity pattern analysis
const calculateActivityPattern = (user: UserData): Promise<number> => {
  // Get hourly message distribution
  const hourlyActivity = await db.messageEvent.groupBy({
    by: ['createdAt'],
    where: { userId: user.discordId },
    _count: true,
  });
  
  // Convert to hour buckets
  const hourBuckets = new Array(24).fill(0);
  hourlyActivity.forEach(item => {
    const hour = new Date(item.createdAt).getHours();
    hourBuckets[hour]++;
  });
  
  // Calculate variance - low variance = suspicious regularity
  const mean = hourBuckets.reduce((a, b) => a + b, 0) / 24;
  const variance = hourBuckets.reduce((sum, val) => sum + Math.pow(val - mean, 2), 0) / 24;
  
  // Low variance = bot-like (too regular)
  // Very high variance = also suspicious (sporadic automated activity)
  if (variance < 5) return 80; // Too regular
  if (variance > 100) return 60; // Too sporadic
  
  return 0;
};
```

### Suspicion Dashboard Component
```tsx
export const SuspicionDetail = ({ userId }: { userId: string }) => {
  const { data: suspicion } = useQuery(['suspicion', userId], () =>
    api.getSuspicionDetail(userId)
  );
  
  if (!suspicion) return <Loading />;
  
  return (
    <div className="space-y-6">
      <div className="flex items-center justify-between">
        <h2 className="text-2xl font-bold">Suspicion Analysis</h2>
        <Badge variant={suspicion.confidence}>
          {suspicion.confidence} confidence
        </Badge>
      </div>
      
      {/* Overall Score */}
      <Card>
        <CardHeader>
          <CardTitle>Overall Suspicion Score</CardTitle>
        </CardHeader>
        <CardContent>
          <div className="flex items-center gap-4">
            <CircularProgress
              value={suspicion.totalScore}
              max={100}
              color={suspicion.totalScore > 75 ? 'red' : suspicion.totalScore > 50 ? 'yellow' : 'green'}
            />
            <div>
              <p className="text-3xl font-bold">{suspicion.totalScore}/100</p>
              <p className="text-muted">{suspicion.recommendedAction}</p>
            </div>
          </div>
        </CardContent>
      </Card>
      
      {/* Contributing Factors */}
      <Card>
        <CardHeader>
          <CardTitle>Contributing Factors</CardTitle>
        </CardHeader>
        <CardContent className="space-y-3">
          {Object.entries(suspicion.factors).map(([key, value]) => (
            <div key={key} className="flex items-center justify-between">
              <span className="capitalize">{key.replace(/([A-Z])/g, ' $1').trim()}</span>
              <div className="flex items-center gap-2">
                <Progress value={value} max={100} className="w-32" />
                <span className="w-12 text-right">{typeof value === 'number' ? Math.round(value) : value}</span>
              </div>
            </div>
          ))}
        </CardContent>
      </Card>
      
      {/* Reasons */}
      <Card>
        <CardHeader>
          <CardTitle>Suspicious Behaviors Detected</CardTitle>
        </CardHeader>
        <CardContent>
          <ul className="space-y-2">
            {suspicion.reasons.map((reason, i) => (
              <li key={i} className="flex items-start gap-2">
                <AlertTriangle className="w-5 h-5 text-warning mt-0.5" />
                <span>{reason}</span>
              </li>
            ))}
          </ul>
        </CardContent>
      </Card>
      
      {/* Actions */}
      <Card>
        <CardHeader>
          <CardTitle>Recommended Actions</CardTitle>
        </CardHeader>
        <CardContent className="space-y-2">
          <Button variant="destructive" className="w-full">
            Ban User
          </Button>
          <Button variant="secondary" className="w-full">
            Add to Watch List
          </Button>
          <Button variant="outline" className="w-full">
            Mark as False Positive
          </Button>
        </CardContent>
      </Card>
    </div>
  );
};
```

## 🎯 Success Criteria

- [x] Suspicion detection accuracy > 85%
- [x] False positive rate < 10%
- [x] Clear explanation for each suspicion score
- [x] Real-time suspicion alerts working
- [x] Admin can provide feedback on accuracy
- [x] Suspicion dashboard is intuitive
- [x] Performance impact minimal

## 📚 Resources

- Anomaly Detection: https://en.wikipedia.org/wiki/Anomaly_detection
- Scikit-learn: https://scikit-learn.org/stable/
- TensorFlow.js: https://www.tensorflow.org/js

## 🔗 Related Issues

- Part of: #5 (Master Roadmap)
- Related: #TBD (Machine Learning for Anomaly Detection)

## ⏱️ Estimated Effort

6-8 days

## 💡 Notes

- Start with rule-based algorithms
- Collect training data for ML models
- Validate with actual suspicious accounts
- Avoid over-fitting to specific patterns
- Allow admin override and feedback
